content/doc/book/security/csrf-protection.adoc

```diff
@@ -0,0 +1,70 @@
+Without protection from CSRF, a Jenkins user or administrator visiting some other web site would allow the operator of that site to perform actions in Jenkins as the victim.
+
+## CSRF Protection in Jenkins
+
+CSRF protection uses a token (called _crumb_ in Jenkins) that is created by Jenkins and sent to the user.
+Any form submissions or similar action resulting in modifications, like triggering builds or changing configuration, requires that the crumb be provided.
+The crumb contains information identifying the user it was created for, so submissions with another user's token would be rejected.
+All of this happens in the background and has no visible impact except in rare circumstances, e.g., after a user's session expired and they logged in again.
+
+NOTE: The documentation on this page applies to Jenkins 2.222 or newer.
+
+
+## Configuring CSRF Protection
+
+In _Manage Jenkins » Configure Global Security » CSRF Protection_, administrators can configure CSRF Protection.
+
+image::managing/configure-global-security-prevent-csrf.png["CSRF Protection in Configure Global Security", role=center]
+
+The _Default Crumb Issuer_ encodes the following information in the https://en.wikipedia.org/wiki/Cryptographic_hash_function[hash] used as crumb:
+
+* The user name that the crumb was generated for
+* The web session ID that the crumb was generated in
+* The IP address of the user that the crumb was generated for
+* A https://en.wikipedia.org/wiki/Salt_(cryptography)[salt] unique to this Jenkins instance
+
+All of this information needs to match when a crumb is sent back to Jenkins for that submission to be considered valid.
+
+The only supported option _Enable proxy compatibility_ removes information about the user IP address from the token.
+This can be useful when Jenkins is running behind a reverse proxy and a user's IP address as seen from Jenkins would regularly change.
+
+NOTE: The web session ID was added in Jenkins 2.176.2 and 2.186 to cause crumb to expire.
+See https://www.jenkins.io/security/advisory/2019-07-17/#SECURITY-626[security advisory] and https://www.jenkins.io/doc/upgrade-guide/2.176/#SECURITY-626[upgrade guide].
+
+Plugins may provide other crumb issuers that use other criteria to determine whether a crumb is valid.
+The plugin:strict-crumb-issuer[Strict Crumb Issuer] provides an alternative crumb issuer implementation that is more customizable.
+
+
+## Working with Scripted Clients
+
+Requests sent using the `POST` method are subject to CSRF protection in Jenkins and generally need to provide a crumb.
+This also applies to scripted clients that **authenticate using username and password**.
+Since the crumb includes the web session ID, clients need to do the following:
+
+* Send a request to the `/crumbIssuer/api` endpoints, requesting a crumb. Note the `Set-Cookie` response header.
+* For all subsequent requests, provide the crumb and the session cookie in addition to username and password.
```

Comment:
Is this requirement to keep the same cookie new? Going to break pretty much every page on the Internet that shows how to send `POST` requests to Jenkins without an API token.

Reply:
New since 2.176.2 (July 2019). Those pages were going to have a bad time anyway since they generally hardcode the CSRF crumb for a given user and never re-request it either.

Reply:
See also https://github.com/jenkinsci/configuration-as-code-plugin/issues/1830.

FWIW I cobbled together something that seems to work long enough to get an API token:

```bash
url=https://jenkins/
user=admin
pass=s3cr3t
crumb=$(curl -s -u $user:$pass -c /tmp/cookies ${url}crumbIssuer/api/xml'?xpath=concat(//crumbRequestField,":",//crumb)')
token=$(curl -s -u $user:$pass -b /tmp/cookies -H $crumb -d newTokenName=general ${url}user/$user/descriptorByName/jenkins.security.ApiTokenProperty/generateNewToken | jq -r .data.tokenValue)
rm /tmp/cookies
```